### DISCUSSION TITLE
GET and PUT route in same route handler causes 405 Method Not Allowed

### DISCUSSION THREAD

**Discussion by chukwumaokere** [vercel/next.js#69331] [App Router]
### Link to the code that reproduces this issue

https://github.com/chukwumaokere/app-route-get-put-bug-minimal-repro

### To Reproduce

1. Clone the repo
2. Start the server locally `pnpm dev`
3. hit the `http://localhost:3000/api/something` route with a `GET` request. See that it works as expected. The request is handled
4. hit the `http://localhost:3000/api/something` route with a `PUT` request. See that it works as expected. The request is handled
5. Run `pnpm run build` to build a production verison
6. Run `pnpm start` to run the production build
7. hit the `http://localhost:3000/api/something` route with a `GET` request. See that it works as expected. The request is handled
8. hit the `http://localhost:3000/api/something` route with a `PUT` request. See that it DOES NOT works as expected. The request is NOT handled. You will receive an HTTP Status 405 Method Not Allowed.

### Current vs. Expected behavior

I expected when I hit the `PUT` endpoint on a production build, even if its next to a `GET` endpoint, it should work. 
Instead I receive a 405 when deployed and when built for production locally. I do NOT get this error when running in dev mode `pnpm dev` so there is no way for me to validate unless I build a production version locally or wait until its deployed.

Here is a screenshot of this running locally on a production build and returning error 405. 
![image](https://github.com/vercel/next.js/assets/16673873/4855801a-bdaf-421a-9ea4-337223206189)

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 10 Pro
  Available memory (MB): 32713
  Available CPU cores: 16
Binaries:
  Node: 22.1.0
  npm: N/A
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.2.3 // Latest available version is detected (14.2.3).
  eslint-config-next: 14.2.3
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.4.5
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Runtime

### Which stage(s) are affected? (Select all that apply)

next build (local), next start (local), Vercel (Deployed)

### Additional context

This issue has been ongoing since Next13 and I've tried researching a resolution but I guess this isn't something people have encountered often since I couldn't find any reports of this issue. It exists into Next 14 and probably will continue to exist. There is a workaround which is just moving the GET or PUT route to its OWN route folder i.e., `/api/something/add` for PUT.

**Comment by chukwumaokere** [comment]
FYI I havent tested this with other routes. Like PUT and POST in the same file, only GET and PUT

**Reply by TeachAcademyDev** [reply]
I also was facing this issue but GET and POST doesnt seem to have it, so weird

**Comment by Firas88Alkhatib** [comment]
opt out of caching seems to solve this.
`export const dynamic = "force-dynamic";`

**Comment by chukwumaokere** [comment]
Can confirm that opting out of caching solves this issue but its not a valid solution in the off chance you want to cache the "cacheable" requests like GET but not cache requests like PUT

Also if nothing else, the error should not be "Method not allowed", this could be a development runtime error with some helpful steps for devs to know this sort of this just isnt supported and they can either move the route, or opt out of caching to resolve it.

**Comment by samcx** [comment]
@chukwumaokere Is there another reason why you would want to have them in the same Route Handler, besides just separating them into two different Route Handlers?

Edit: you should be able to do this actually, we're just not handling the edge cases with `force-static`

**Reply by TeachAcademyDev** [reply]
i believe one major issue bothering us is why not throw some errors in dev environment so we can be wary of it? We are in 2024 with major advancements in technology and i believe now should be the time where our dev environments catch up to our production environments as much as it can and not rely on waiting for the unknown to happen when the app hits production. I am talking in terms of such bugs like it makes zero sense for such an error. Who knows what other errors is lurking there and what if it crashes many websites?

**Reply by samcx** [reply]
@TeachAcademyDev I don't disagree with you—we could be handling this with better error handling or warning messages!

> Who knows what other errors is lurking there and what if it crashes many websites?

We we have 6500+ tests and are pushing up bug fixes, improvements everyday.  If there was a severe bug breaking many sites, we would notice it asap and there is no doubt it will get patched very quickly.

**Reply by samcx** [reply]
@chukwumaokere @TeachAcademyDev :dug-deep: into this more!

So my apologies—we _should_ be able to use multiple request methods in Route Handlers.

- The edge case at hand we need to fix is throw a warning/error with static Route Handlers. When a Route Handler is static, only `GET`, `HEAD`, and `OPTIONS` can be used.
- Update the documentation to reflect this.

Let me know if that clarifies!

**Reply by Tahmeed-programmer** [reply]
How do you determine a route handler is static? and yes please update the docs , much appreciated!

**Reply by samcx** [reply]
@Tahmeed-programmer Route Handlers were static by default before v15 → https://nextjs.org/blog/next-15-rc#get-route-handlers-are-no-longer-cached-by-default. Now, with v15 and later (the newest canaries), you don't need to worry about `export const dynamic = "force-dynamic";`—it is always dynamic by default. If you want it to be static (with v15 and later), you'll need to use `export const dynamic = "force-static";` instead.

Essentially, the opposite is true pre-v15 (static by default).

Unfortunately, we do not have descriptive enough logs to show whether a Route Handler is static or not. But, when deployed (on Vercel for example), the `X-Vercel-Cache` response header will always be a `MISS` for dynamic Route Handlers.

![CleanShot 2024-08-29 at 12 35 22@2x](https://github.com/user-attachments/assets/509611d0-7e78-4548-8895-b59753d94a03)

**Reply by Tahmeed-programmer** [reply]
I tried using next v15 2 months ago i think and i was getting some weird errors with regards to next auth and some libraries, when can we use it for production? I really want to use v15. Furthermore when a route handler is static , does it mean every time i sent a request with same params it caches the response and returns the cached response back to me? any idea on how long it caches?

**Reply by samcx** [reply]
> i was getting some weird errors with regards to next auth and some libraries, when can we use it for production

I'm not sure where these libraries stand with v15—we would have to take a look at these errors to see if it's because the library is not ready yet with v15.

> does it mean every time i sent a request with same params it caches the response and returns the cached response back to me?

Yes! You can set a revalidate → https://nextjs.org/docs/app/building-your-application/routing/route-handlers#revalidating-cached-data. The cached response will last essentially forever (always get a `HIT`).

**Reply by Tahmeed-programmer** [reply]
So if a route has both GET and POST handlers , will the GET handler responses be cached? if you want caching for the GET requests then how would you personally structure the api routes since usually we host both GET and POST in a single route.ts file. So for instnace getting list of appointments and posting a new appointment is in the same route.ts file under the same folder called /appointments, how would you structure the files if you wanted the GET handler responses to be cached?

**Reply by samcx** [reply]
> So if a route has both GET and POST handlers , will the GET handler responses be cached?

POST will never get cached (I think our documentation needs to be clearer about this), so depending on the version, your GET can be cached or not (v14 and previous will be static by default, so cached, v15 and beyond is dynamic by default), or if you put a `export const dynamic = 'force-static'

Personally, I would just separate GET and POST Route Handlers to separate files, but it's fine if they're in the same file.

**Reply by Tahmeed-programmer** [reply]
I see, because what i though was if you have a post handler inside a route.ts file with a get handler then that GET handler is not cached by default. Correct me if i am wrong? yeah the documentation is a bit confusing and if it was me what i would have done is broke it down into all possible cases i can think of in a faq way like for sure i can help you all no issues. Nextjs documentation is really good but can be improved more. You and your team hard work is appreciated despite some of the critics of nextjs . Those who criticize nextjs and other frameworks in general are more than welcome to code in pure JavaScript and worry about the tons of headaches that come with it such as optimizing images, caching responses etc like wow i don't think many new devs realize how much of a powerhouse tools like nextjs is , and this is from someone who coded in vanilla js , vanilla css as well as php in my early programming journey. Frameworks are there for a reason. Thank you for been responsive to my queries and developing such an amazing framework.

**Reply by samcx** [reply]
> if you have a post handler inside a route.ts file with a get handler then that GET handler is not cached by default.

They're independent of each other, so it just depends what version you're using, or if you're using a Route Segment Config to explicitly state the state of the cache.

> Nextjs documentation is really good but can be improved more

Totally agree with this.

I actually went ahead and updated the documentation to mention this correctly! → https://nextjs.org/docs/app/building-your-application/routing/route-handlers#caching

**Reply by Tahmeed-programmer** [reply]
so if am not mistaken since me and many others still use nextjs 14 , we have to be aware that a static route handler cannot have post and put requests at the same route.ts file, correct? for me i never had an issue with post and get been in the same file. the issue i saw was with PUT and GET been together.

**Reply by samcx** [reply]
we have to be aware that a static route handler cannot have post and put requests at the same route.ts file, correct?

> The caching only applies to `GET`.

> the issue i saw was with PUT and GET been together.

I agree those two together need to get fixed!

**Reply by Tahmeed-programmer** [reply]
@samcx So i have discovered some major errors with nextjs15 itself which is first and foremost that if lets say you have a classname string that spans multiple lines it throws an error and this is problematic for many of us and apparently no codemod available to tackle it. So the code   "section className="w-full bg-white overflow-hidden 
    py-12 md:py-28"  
throws the error -
 "Module parse failed: Unterminated string constant (175:50)
|     }
|     return /*#__PURE__*/ _jsxDEV("section", {
>         className: "jsx-33324318e7668a2d" + " " + "w-full bg-white overflow-hidden
|     py-12 md:py-28",
|         children: [" and to fix it you have to make the whole classname in a single line. Thats one error , second error is sinc nextjs 15 uses await for pages ,router params ,cookies etc this is causing issues with next-auth v4 as well as 3rd party services like zego-cloud react package , sentry packages etc as those pages doesnt use await for cookies or headers and the like. So can you all maybe allow nextjs15 to overlook it instead of just forcing the packages to not work just because they didnt adapt to the async model of nextjs as there are so many many pacakges that dont use async for await as its a new architecture. So please something needs to be done about this rather than forcing many such packages to adapt as they are so many in magnitude. i for now will try to find a nextjs14 version that supports partial pre-rendering. Actually can u tell me a version?

Here are some of the other errors - 

Error: Route "/api/auth/[...nextauth]" used `cookies().getAll()`. `cookies()` should be awaited before using its value. Learn more: https://nextjs.org/docs/messages/sync-dynamic-apis
    at createCookiesAccessError (webpack-internal:///(rsc)/../../node_modules/.pnpm/next@15.0.1_@babel+core@7.24.5_@opentelemetry+api@1.9.0_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/request/cookies.js:381:12)
    at logDedupedError (webpack-internal:///(rsc)/../../node_modules/.pnpm/next@15.0.1_@babel+core@7.24.5_@opentelemetry+api@1.9.0_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/create-deduped-by-callsite-server-error-loger.js:70:25)
    at syncIODev (webpack-internal:///(rsc)/../../node_modules/.pnpm/next@15.0.1_@babel+core@7.24.5_@opentelemetry+api@1.9.0_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/request/cookies.js:375:5)

**Comment by samcx** [comment]
@chukwumaokere I am not longer able to replicate this error on the latest stable or canary.

**Reply by Tahmeed-programmer** [reply]
fantastic. Just a quick question, since i have upgraded to nextjs15 and react 19, do you know any codemod available for fixing the tyepscript breaking changes such as this error -"Type 'ForwardRefExoticComponent<DialogContentProps & RefAttributes<HTMLDivElement>>' does not satisfy the constraint 'ElementType<any, keyof IntrinsicElements>'.
  Type 'ForwardRefExoticComponent<DialogContentProps & RefAttributes<HTMLDivElement>>' is not assignable to type 'FunctionComponent<any>'.
    Type 'import("C:/Users/alamt/Desktop/Teach_Academy_Monorepo/node_modules/.pnpm/types-react@19.0.0-rc.1/node_modules/types-react/index").ReactNode' is not assignable to type 'React.ReactNode'."

**Reply by samcx** [reply]
@Tahmeed-programmer Not sure to be honest! We've been working on Next.js specific codemods, but nothing specific to TypeScript and React 19.

**Reply by Tahmeed-programmer** [reply]
i honestly kinda gave up after facing the amount of bugs since upgrading to nextjs15 such as unterminated string constant literals which is coming from ract 19 typescript changes, then some weird "use context" function typescript error and so forth. I share the opinion of many others on why did nextjs release version 15 when react 19 is not stable? already people have found bugs which can be tracked in github for nextjs15. At least the docs should have attempted to tell us about potential errors and how to resolve them when using react 19 unstable version. I am back to nextjs14 for now and hopefully one day a lot of us can use nextjs15. Is it because of the upcoming conference the team decided to release and mark the latest version as stable? if nextjs15 could work with react 18 then it would be great but unfortunately too many issues from react 19

**Reply by samcx** [reply]
@Tahmeed-programmer Tough to say if this is because React 19 is "unstable," or if these are just breaking changes that need to be addressed.

Is it possible to share a reproduction that creates these TypeScript errors?

**Reply by Tahmeed-programmer** [reply]
I cannot share the repo due to company policy as its a private repo but in general there are lots of breaking changes especially starting with headless-ui incompatibilty with react-19 and there are more as well. People have alerted the creator of shadcn as well with regards to typescript errors of the components. I can share with you a link of someone who talked about headless-ui breaking changes - https://github.com/tailwindlabs/headlessui/issues/3538

**Reply by Tahmeed-programmer** [reply]
@samcx So i have discovered some major errors with nextjs15 itself which is first and foremost that if lets say you have a classname string that spans multiple lines it throws an error and this is problematic for many of us and apparently no codemod available to tackle it. So the code   "section className="w-full bg-white overflow-hidden 
    py-12 md:py-28"  
throws the error -
 "Module parse failed: Unterminated string constant (175:50)
|     }
|     return /*#__PURE__*/ _jsxDEV("section", {
>         className: "jsx-33324318e7668a2d" + " " + "w-full bg-white overflow-hidden
|     py-12 md:py-28",
|         children: [" and to fix it you have to make the whole classname in a single line. Thats one error , second error is sinc nextjs 15 uses await for pages ,router params ,cookies etc this is causing issues with next-auth v4 as well as 3rd party services like zego-cloud react package , sentry packages etc as those pages doesnt use await for cookies or headers and the like. So can you all maybe allow nextjs15 to overlook it instead of just forcing the packages to not work just because they didnt adapt to the async model of nextjs as there are so many many pacakges that dont use async for await as its a new architecture. So please something needs to be done about this rather than forcing many such packages to adapt as they are so many in magnitude. i for now will try to find a nextjs14 version that supports partial pre-rendering. Actually can u tell me a version?

Here are some of the other errors - 

Error: Route "/api/auth/[...nextauth]" used `cookies().getAll()`. `cookies()` should be awaited before using its value. Learn more: https://nextjs.org/docs/messages/sync-dynamic-apis
    at createCookiesAccessError (webpack-internal:///(rsc)/../../node_modules/.pnpm/next@15.0.1_@babel+core@7.24.5_@opentelemetry+api@1.9.0_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/request/cookies.js:381:12)
    at logDedupedError (webpack-internal:///(rsc)/../../node_modules/.pnpm/next@15.0.1_@babel+core@7.24.5_@opentelemetry+api@1.9.0_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/create-deduped-by-callsite-server-error-loger.js:70:25)
    at syncIODev (webpack-internal:///(rsc)/../../node_modules/.pnpm/next@15.0.1_@babel+core@7.24.5_@opentelemetry+api@1.9.0_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/request/cookies.js:375:5)
